python/ray/_private/package.py

```diff
@@ -0,0 +1,276 @@
+import hashlib
+import logging
+import inspect
+
+from filelock import FileLock
+from pathlib import Path
+from zipfile import ZipFile
+from ray.job_config import JobConfig
+from enum import Enum
+from ray.experimental import internal_kv
+from ray.core.generated.common_pb2 import RuntimeEnv
+from urllib.parse import urlparse
+import os
+import sys
+
+PKG_DIR = "/tmp/ray/runtime_resources"
+Path(PKG_DIR).mkdir(parents=True, exist_ok=True)
+
+logger = logging.getLogger(__name__)
+
+FILE_SIZE_WARNING = 10 * 1024 * 1024  # 10MB
+FILE_SIZE_LIMIT = 50 * 1024 * 1024  # 50MB
+
+
+class Protocol(Enum):
+    """A enum for supported backend storage."""
+
+    # For docstring
+    def __new__(cls, value, doc=None):
+        self = object.__new__(cls)
+        self._value_ = value
+        if doc is not None:
+            self.__doc__ = doc
+        return self
+
+    GCS = "gcs", "For package created and managed by system."
+    PIN_GCS = "pingcs", "For package created and managed by users."
+
+
+def _xor_bytes(left, right):
+    if left and right:
+        return bytes(a ^ b for (a, b) in zip(left, right))
+    return left or right
+
+
+def _zip_module(path, relative_path, zip_handler):
+    """Go through all files and zip them into a zip file"""
+    for from_file_name in path.glob("**/*"):
+        file_size = from_file_name.stat().st_size
+        if file_size >= FILE_SIZE_LIMIT:
+            raise RuntimeError(f"File {from_file_name} is too big, "
+                               "which currently is not allowd ")
+        if file_size >= FILE_SIZE_WARNING:
+            logger.warning(
+                f"File {from_file_name} is too big ({file_size} bytes). "
+                "Consider exclude this file in working directory.")
+        to_file_name = from_file_name.relative_to(relative_path)
+        zip_handler.write(from_file_name, to_file_name)
+
+
+def _hash_modules(path):
+    """
+    Helper function to create hash of a directory.
+    It'll go through all the files in the directory and xor
+    hash(file_name, file_content) to create a hash value.
+    """
+    hash_val = None
+    BUF_SIZE = 4096 * 1024
+    for from_file_name in path.glob("**/*"):
+        md5 = hashlib.md5()
+        md5.update(str(from_file_name).encode())
+        if not Path(from_file_name).is_dir():
+            with open(from_file_name, mode="rb") as f:
+                data = f.read(BUF_SIZE)
+                if not data:
+                    break
+                md5.update(data)
+        hash_val = _xor_bytes(hash_val, md5.digest())
+    return hash_val
+
+
+def _get_local_path(pkg_uri) -> str:
+    (_, pkg_name) = _parse_uri(pkg_uri)
+    return os.path.join(PKG_DIR, pkg_name)
+
+
+def _parse_uri(pkg_uri: str) -> (Protocol, str):
+    uri = urlparse(pkg_uri)
+    protocol = Protocol(uri.scheme)
+    return (protocol, uri.netloc)
+
+
+# TODO(yic): Fix this later to handle big directories in better way
+def get_project_package_name(working_dir, modules) -> str:
```

Comment:
```suggestion
def get_project_package_name(working_dir: str, modules: List[str]) -> str:
```

Please add types for all new code (some old code isn't typed, but we should add them as possible)